src/etwlogger_driver.c

```diff
@@ -277,50 +145,61 @@ static void perform_EventWriteLogVerboseEvent(const char* message, const char* f
     );
 
 #if CALL_CONSOLE_LOGGER
-    consolelogger_log(AZ_LOG_VERBOSE, NULL, NULL, 0, LOG_LINE, "%s", message);
+    consolelogger_log(AZ_LOG_VERBOSE, file, func, line, LOG_LINE, "%s", message);
 #endif
 }
 
+
+static const char vsnprintf_failure_message[] = "failure in vsnprintf";
+static const char FormatMessageA_failure_message[] = "failure in FormatMessageA";
 void etwlogger_log_with_GetLastError(const char* file, const char* func, int line, const char* format, ...)
 {
     DWORD lastError;
-    char* lastErrorAsString;
 
     lastError = GetLastError();
     lazyRegisterEventProvider();
 
     va_list args;
     va_start(args, format);
 
-    lastErrorAsString = lastErrorToString(lastError);
-    if (lastErrorAsString == NULL)
-    {
-        char* userMessage = vprintf_alloc(format, args);
-        if (userMessage == NULL)
+    char message[LOG_SIZE_REGULAR * (LOG_SIZE_REGULAR>=sizeof(vsnprintf_failure_message))]; /*this construct will generate a compile time error (array of size 0) when LOG_SIZE_REGULAR is not enough to hold even the failure message*/
+    {/*scope for constructing the user (format,...)*/
+        int vsnprintf_result;
+        vsnprintf_result = vsnprintf(message, sizeof(message), format, args);
+        if (
+            (vsnprintf_result < 0)||
+            (vsnprintf_result >= sizeof(message))
+            )
         {
-            perform_EventWriteLogLastError("unable to print user error", file, func, line, "last error was erroneously NULL");
+            (void)memcpy(message, vsnprintf_failure_message, sizeof(vsnprintf_failure_message));
         }
         else
         {
-            perform_EventWriteLogLastError(userMessage, file, func, line, "last error was erroneously NULL");
-            free(userMessage);
+            /*all fine, message now contains user message*/
         }
     }
+
+    char lastErrorAsString[MESSAGE_BUFFER_SIZE *(MESSAGE_BUFFER_SIZE>sizeof(FormatMessageA_failure_message))];
+    if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, lastError, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), lastErrorAsString, sizeof(lastErrorAsString), NULL) == 0)
+    {
+        (void)memcpy(lastErrorAsString, FormatMessageA_failure_message, sizeof(FormatMessageA_failure_message));
+    }
     else
     {
-        char* userMessage = vprintf_alloc(format, args);
-        if (userMessage == NULL)
+        /*remove extraneous newlines from FormatMessageA's output*/
+        char* whereAreThey;
+        if ((whereAreThey = strchr(lastErrorAsString, '\r')) != NULL)
```

Comment:
Just feeling like consolelogger.c/etwlogger_driver.c have a lot of common code, maybe a shared helper for "strip newlines" is one example of code that could be shared. /consider

Reply:
Yeah I agree, I think re-organizing logging to be more structured a definitely a task, I would like to track it with a different Pull Request.

---
In reply to: [453752517](https://github.com/Azure/azure-c-logging/pull/11#discussion_r453752517) [](ancestors = 453752517)